Describe the semi-automated procedure that migrates Discovery Service V1 to V2. (#1861)
neo4j/docs-operations: modules/ROOT/pages/clustering/setup/discovery.adoc

It is important that both `dbms.kubernetes.service_port_name` and `dbms.kubernetes.discovery.v2.service_port_name` are set during the operation.
For more information, see <<clustering-discovery-k8s>>.

[[discovery-v1-to-v2-semi-automated-procedure]]

> === Moving whole cluster with a procedure

Reviewer: ```suggestion
=== Procedure for moving an entire cluster
```

Reviewer: To align the style of the first and second titles, we can make them: 

- Moving an entire cluster with a procedure

- Running a procedure on each server

or 

- Procedure for moving an entire cluster

- Per server procedure

I'd prefer the second variant.

Author: I have changed the titles to align with the second option.

> The simplest way to migrate from Discovery Service `V1` to `V2` is to use a semi-automated procedure, which simplifies the migration process.

Reviewer: We try not to evaluate if this or that approach is simple or difficult.

dbms.cluster.discovery.endpoints=server01.example.com:5000,server02.example.com:5000,server03.example.com:5000
dbms.cluster.discovery.v2.endpoints=server01.example.com:6000,server02.example.com:6000,server03.example.com:6000
----
+
. In Cypher Shell, connect to the `system` database of any server
+
[source, shell, role=nocopy noplay]
----
./cypher-shell -a bolt://localhost:7681 -d system
----
. Run the procedure:
+
[source,cypher]
----
CALL dbms.cluster.showParallelDiscoveryState();
----
+

> The output indicates mode `V1_ONLY`, i.e., only `V1` is running on this server.

Reviewer: 'on this server' or 'on this cluster'?
You say 'on this server' here but the procedure is for moving the entire cluster.

Author: This procedure `CALL dbms.cluster.showParallelDiscoveryState();` is used to check the state of a single member. So we need to use the term **server**.  This procedure `CALL internal.dbms.cluster.moveToNextDiscoveryVersion()` moves several members to the next discovery version

Author: I think your question is: why do we query the discovery state (CALL dbms.cluster.showParallelDiscoveryState()) of a single server after calling CALL internal.dbms.cluster.moveToNextDiscoveryVersion(); instead of checking the state of other servers?

Our goal is to minimize the number of steps the DB admin needs to take to migrate from V1_ONLY to V2_ONLY. We are quite confident that if the state of a single server matches (i.e., showParallelDiscoveryState returns 'Matching'), the states of the other servers are also likely to match. This is the reason why we don't ask db-admins to check the state of other servers.

+
.Expected result
[queryresult]
----
+-------------------------------------------------------------+
| mode      | stateComparison | v1ServerCount | v2ServerCount |
+-------------------------------------------------------------+
| "V1_ONLY" | "N/A"           | "3"           | "N/A"         |
+-------------------------------------------------------------+
----
. Run the following procedure to turn on `V2` in the background for all servers, but keep `V1` running in the foreground.
If the procedure fails, read the error message and resolve the issue manually.

> One possible reason for failure is that some servers might have had inconsistent xref:configuration/configuration.settings.adoc#config_dbms.cluster.discovery.version[`dbms.cluster.discovery.version`] settings initially.

Reviewer: ```suggestion
One possible reason for failure is that some servers might have had inconsistent xref:configuration/configuration-settings.adoc#config_dbms.cluster.discovery.version[`dbms.cluster.discovery.version`] settings initially.
```

+
[source,cypher]
----

> CALL internal.dbms.cluster.moveToNextDiscoveryVersion();

Author: done

----
. Check the state again:
+
[source,cypher]
----
CALL dbms.cluster.showParallelDiscoveryState();
----
+

> Now the returned mode for this server must be `V1_OVER_V2` and the `stateComparison` must show that the states are matching.

Reviewer: once again - 'for this server'. What's about all other servers?

Author: https://github.com/neo4j/docs-operations/pull/1861#discussion_r1809060567

If they are not, wait and try again till matching.
+
.Expected result
[queryresult]
----
+----------------------------------------------------------------+
| mode         | stateComparison | v1ServerCount | v2ServerCount |
+----------------------------------------------------------------+
| "V1_OVER_V2" | "Matching"      | "3"           | "3"           |
+----------------------------------------------------------------+
----
. Run the once again the following procedure to turn on `V1` in the background for all servers, but keep `V2` running in the foreground.
+
[source,cypher]
----

> CALL internal.dbms.cluster.moveToNextDiscoveryVersion();

Reviewer: ```suggestion
CALL dbms.cluster.moveToNextDiscoveryVersion();
```

----
. Check that transition to `V2_OVER_V1` was successful:
+
[source,cypher]
----
CALL dbms.cluster.showParallelDiscoveryState();
----
+

> Now the returned mode for this server must be `V2_OVER_V1` and the `stateComparison` must show that the states are matching.

Reviewer: same here - 'the returned mode for this server...' or it should be - 'for the cluster...'?

Author: https://github.com/neo4j/docs-operations/pull/1861#discussion_r1809060567

+----------------------------------------------------------------+
| mode         | stateComparison | v1ServerCount | v2ServerCount |
+----------------------------------------------------------------+
| "V2_OVER_V1" | "Matching"      | "3"           | "3"           |
+----------------------------------------------------------------+
----
. Finally, turn off `V1` by running the following procedure:
+
[source,cypher]
----

> CALL internal.dbms.cluster.moveToNextDiscoveryVersion();

Reviewer: ```suggestion
CALL dbms.cluster.moveToNextDiscoveryVersion();
```

Author: done

| "V2_ONLY" | "N/A"           | "N/A"         | "3"           |
+-------------------------------------------------------------+
----
+
.Important
[IMPORTANT]
====
Remember to update the _neo4j.conf_ files for all the servers.
The switching using procedures does not persist anything to disk.
Therefore, when a server restarts, it starts right back with only v1 running.
As such, ensure that `dbms.cluster.discovery.version=V2_ONLY`, and that `dbms.cluster.discovery.v2.endpoints` or `dbms.kubernetes.discovery.v2.service_port_name`
are set as required, so that the servers start with v2 running on the next restart.
====

[[discovery-v1-to-v2-procedures]]

> === Per server procedure

Reviewer: ```suggestion
=== Running a procedure on each server
```

Reviewer: see comment in the beginning

// By using just procedures, the nice thing is that during the user's normal version upgrade, they can also add the new settings required for v2 discovery service.
// Then, when they are ready to migrate to v2, they can use the procedures and then finally, set the version to 'V2' in the neo4j.conf at the very end - which can just sit there till the next restart
// so no lighthouse-specific server restarts are required if all is done correctly.
//Note that the settings detailed above must first be set and the servers restarted to allow the settings to take effect.

[IMPORTANT]
====

> If the cluster procedure had problems, or you need to change individual servers to resolve an issue, this approach will let you do so.

Reviewer: ```suggestion
If the semi-automated procedure fails, or you need to change individual servers to resolve an issue, this approach will let you do so.
```

+-------------------------------------------------------------+
----

> [[discovery-v1-to-v2-semi-automated-procedure]]

Reviewer 2: Also, there is a line just before 'Preparation' that says
> There are three ways to move
There are now four.

are set as required, so that the servers start with v2 running on the next restart.
====

[[discovery-v1-to-v2-semi-automated-procedure]]
=== Semi-automated procedure

Another way to migrate from Discovery Service `V1` to `V2` is to use a semi-automated procedure, which simplifies the migration process
It automates the approach described in the xref:clustering/setup/discovery.adoc#_using_procedures[previous section].

> . Make sure that server side routing is enabled as detailed in xref:https://neo4j.com/docs/operations-manual/current/clustering/setup/routing/#clustering-routing[Server-side routing] section

Reviewer: ```suggestion
. Make sure that server side routing is enabled as detailed in xref:clustering/setup/routing.adoc#clustering-routing[Server-side routing] section
```

are set as required, so that the servers start with v2 running on the next restart.
====

[[discovery-v1-to-v2-semi-automated-procedure]]

> === Semi-automated procedure

Reviewer 3: Might be worth renaming the current `Using procedures` to something like `Manual procedures`, but leave it at the bottom of the page as it is currently

> === Semi-automated procedure

Reviewer 2: The order of sections should be:

- Preparation
- Semi-automated procedure
- Manual procedures
- In-place rolling
- New server rolling

I.e. this new procedure should be the default option for someone reading the docs.

And maybe we can call the new section 'Moving whole cluster with a procedure', and what is now 'Manual procedures' is 'Per server procedure', and has a note at the beginning like
> If the cluster procedure had problems, or you need to change individual servers to resolve an issue, this approach will let you do so.

> Another way to migrate from Discovery Service `V1` to `V2` is to use a semi-automated procedure, which simplifies the migration process

Reviewer 2: Missing full stop
```suggestion
Another way to migrate from Discovery Service `V1` to `V2` is to use a semi-automated procedure, which simplifies the migration process.
```

It automates the approach described in the xref:clustering/setup/discovery.adoc#discovery-v1-to-v2-procedures[previous section].

. Make sure that server side routing is enabled as detailed in xref:clustering/setup/routing.adoc#clustering-routing[Server-side routing] section
. For all the servers, ensure that new settings are updated to the configuration as detailed the xref:clustering/setup/discovery.adoc#_preparation[Preparation] section.

> . Connect to any server using server-side routing

Reviewer 2: We are not connecting 'with server side routing'. I don't believe we even need to connect with `neo4j://`, because it's not a routed query. So I would just say:
```suggestion
. Connect to any server.
```

+
[source,cypher]
----
./cypher-shell -a neo4j://localhost:7691 -d system
----

> . Run the following procedure three times to move all members from `V1_ONLY` to `V2_ONLY`. If the procedure fails, read the error message and resolve the issue manually.

Reviewer 2: This needs more detail. We still need the user to for instance check the difference score so they wait for some level of convergence.

Author: We discussed this verbally, so we decided that it is enough to call `showParallelDiscoveryState()` on a single member.

One possible reason for failure is that some members might have had inconsistent `dbms.cluster.discovery.version` settings initially.
Refer to the previous section for guidance on debugging and fixing issues.
+
[source,cypher]
----

> CALL internal.dbms.cluster.moveToNextDiscoveryVersion();

Reviewer 2: When we publish these docs, the procedure will not have the `internal.` prefix.

----
. To check that the migration was successful, you can execute `CALL dbms.cluster.showParallelDiscoveryState()` on all members of the cluster. You should see that V2_ONLY is the current version.
+
[queryresult]
----
+-------------------------------------------------------------+
| mode      | stateComparison | v1ServerCount | v2ServerCount |
+-------------------------------------------------------------+
| "V2_ONLY" | "N/A"           | "N/A"         | "3"           |
+-------------------------------------------------------------+
----

> . In the last step, set the `dbms.cluster.discovery.version=V2_ONLY` for each member of the cluster, so that the version change is retained after a restart.

Reviewer 2: We need more detail on what 'set' means here, i.e.
```suggestion
. In the last step, add `dbms.cluster.discovery.version=V2_ONLY` to the config file for each member of the cluster, so that the version change is retained after a restart.
```
